File: agv_main/src/agv_main.cpp
#include "ros/ros.h"
#include <agv_main/agv_main.h>
#include <agv_main/agv_action.h>
// #include <agv_main/lift_up.h>s
#include <string>

#include <geometry_msgs/Twist.h>
#include "geometry_msgs/PoseStamped.h"


// bool lift_up_srv(agv_main::lift_up::Request  &req, agv_main::lift_up::Response &res)
// {
//   ROS_INFO("agv_main.cpp-7-lift_up_srv()");
//   action_.header.stamp = ros::Time::now();
//   action_.action = 5;
//   action_.status = 1;
//   PublishStatusService(action_);

//   res.sum = req.a + req.b;
//   ROS_INFO("agv_main.cpp-10-request: x=%ld, y=%ld", (long int)req.a, (long int)req.b);
//   ROS_INFO("agv_main.cpp-11-sending back response: [%ld]", (long int)res.sum);
//   return true;
// }
// bool lift_down_srv(agv_main::lift_up::Request  &req, agv_main::lift_up::Response &res)
// {
//   ROS_INFO("agv_main.cpp-16-lift_down_srv()");
//   return true;
// }
// bool charging_in_srv(agv_main::lift_up::Request  &req, agv_main::lift_up::Response &res)
// {
//   ROS_INFO("agv_main.cpp-21-charging_in_srv()");
//   return true;
// }
// bool charging_out_srv(agv_main::lift_up::Request  &req, agv_main::lift_up::Response &res)
// {
//   ROS_INFO("agv_main.cpp-26-charging_out_srv()");
//   return true;
// }

// void PublishStatusService(agv_main::agv_action action){
//   status_srv_pub_.publish(action);
//   ROS_INFO("agv_main.cpp-13-publish status action lift up!");
// }

void decodeJson(std::string type, std::string data){
  ROS_INFO("agv_main.cpp-59-type: %s", type.c_str());
  ROS_INFO("agv_main.cpp-60-data: %s", data.c_str());
}
void initialAgvAction(const agv_main::agv_action::ConstPtr& msg)
{
  ROS_INFO("agv_main.cpp-50-initialAgvAction()");
  uint8_t action, status;
  action = msg->action;
  status = msg->status;
  std::string type = msg->type;
  std::string data = msg->data;

  // std::string type = "geometry_msgs/PoseStamped";
  // std::string data = "{\"header":{"seq":1,"stamp":0,"frame_id":"map"},"pose":{"position":{"x":0,"y":0,"z":0},"orientation":{"x":0,"y":0,"z":0,"w":0}}}\";

  ROS_INFO("agv_main.cpp-57-action: %d", action);
  ROS_INFO("agv_main.cpp-58-status: %d", status);

  action_state = ActionState(action);
  switch(action_state){
    case ACTION_FLOAT:
      break;
    case ACTION_MANUAL:
      break;
    case ACTION_MOVE_BASE:
      decodeJson(type, data);
      break;
    case ACTION_INITIAL_POSE:
      break;
    case ACTION_CHARGING_IN:
      break;
    case ACTION_CHARGING_OUT:
      break;
    case ACTION_LIFT_UP:
      break;
    case ACTION_LIFT_DOWN:
      break;
    default:
    {}
  }
}

int main(int argc, char **argv)
{
  ROS_INFO("agv_main.cpp-41-main()");
  ros::init(argc, argv, "agv_main");

  // ros::ServiceServer lift_up_service_ = n.advertiseService("lift_up_", lift_up_srv);
  // ROS_INFO("agv_main.cpp-36-Advertise service: lift_up_");
  // ros::ServiceServer lift_down_service_ = n.advertiseService("lift_down_", lift_down_srv);
  // ROS_INFO("agv_main.cpp-38-Advertise service: lift_down_");
  // ros::ServiceServer chaging_in_service = n.advertiseService("charging_in_", charging_in_srv);
  // ROS_INFO("agv_main.cpp-40-Advertise service: chaging_in_");
  // ros::ServiceServer charging_out_service = n.advertiseService("charging_out_", charging_out_srv);
  // ROS_INFO("agv_main.cpp-42-Advertise service: chaging_out_");

  // // Publish the topic /current_pose
  // status_srv_pub_ = n.advertise<agv_main::agv_action>("agv_service_status", 0);
  // ROS_INFO("move_base.cpp-93-Publish topic: /agv_service_status");
  
  ros::NodeHandle n;
  action_state = ACTION_FLOAT;

  ros::Subscriber agv_action_ = n.subscribe("agv_action", 2, &initialAgvAction);
  ROS_INFO("agv_main.cpp-85-Subscriber topic: /agv_action");

  ros::spin();
  return 0;
}

File: agv_main/src/agv_main.py
#!/usr/bin/env python
import rospy
from rospy_message_converter import json_message_converter
from std_msgs.msg import String
from geometry_msgs.msg import PoseStamped
from geometry_msgs.msg import Twist
from geometry_msgs.msg import PoseWithCovarianceStamped

# from agv_main.msg import agv_action
from rospy_message_converter.msg import agv_action


pub_init_pose = rospy.Publisher('initialpose', PoseWithCovarianceStamped, queue_size=10)
rospy.loginfo("agv_main.py-Publisher topic /initialpose")
pub_cmd_vel = rospy.Publisher('cmd_vel', Twist, queue_size=10)
rospy.loginfo("agv_main.py-Publisher topic /cmd_vel")
pub_charging = rospy.Publisher('charging_action', agv_action, queue_size=10)
rospy.loginfo("agv_main.py-Publisher topic /charging_action")
pub_lift = rospy.Publisher('lift_action', agv_action, queue_size=10)
rospy.loginfo("agv_main.py-Publisher topic /lift_action")
pub_move_base_goal = rospy.Publisher('move_base_simple/goal', PoseStamped, queue_size=10)
rospy.loginfo("agv_main.py-Publisher topic /move_base_simple/goal")

def actionMode(argument):
    switcher = {
        0: "ACTION_FLOAT",
        1: "ACTION_MANUAL",
        2: "ACTION_MOVE_BASE",
        3: "ACTION_INITIAL_POSE",
        4: "ACTION_CHARGING_IN",
        5: "ACTION_CHARGING_OUT",
        6: "ACTION_LIFT_IN",
        7: "ACTION_LIFT_UP",
        8: "ACTION_LIFT_DOWN",
        9: "ACTION_LIFT_OUT"
    }
    return switcher.get(argument, "Invalid action")

def initialAgvAction(msg):
    rospy.loginfo("agv_main.py-initialAgvAction()")   
    # typeData = "geometry_msgs/PoseStamped"
    # message = PoseStamped({"header": {"stamp": {"secs": 0, "nsecs": 0}, "frame_id": "", "seq": 1}, "pose": {"position": {"y": 0.0, "x": 0.0, "z": 0.0}, "orientation": {"y": 0.0, "x": 0.0, "z": 0.0, "w": 0.0}}})
    # data = json_message_converter.convert_ros_message_to_json(message)

    action = msg.action
    typeData = msg.type
    data = msg.data
    rospy.loginfo("agv_main.py-data: " + str(data))
    message = json_message_converter.convert_json_to_ros_message(typeData, data)
    rospy.loginfo("agv_main.py-type: %s", typeData)
    # rospy.loginfo("agv_main.py-message: %s", message)
    # rospy.loginfo("agv_main.py-action: " + str(action))
    action_mode = actionMode(action)
    rospy.loginfo("agv_main.py-action_mode: " + str(action_mode))
    if (action_mode == "ACTION_FLOAT"):
        floatFunction()
    elif (action_mode == "ACTION_MANUAL"):
        manualFunction(message)
    elif (action_mode == "ACTION_MOVE_BASE"):
        navigationFunction(message)
    elif (action_mode == "ACTION_INITIAL_POSE"):
        initPoseFunction(message)
    elif (action_mode == "ACTION_CHARGING_IN"):
        chargingInFunction(msg)
    elif (action_mode == "ACTION_CHARGING_OUT"):
        chargingOutFunction(msg)
    elif (action_mode == "ACTION_LIFT_IN"):
        liftInFunction(msg)
    elif (action_mode == "ACTION_LIFT_UP"):
        liftUpFunction(msg)
    elif (action_mode == "ACTION_LIFT_DOWN"):
        liftDownFunction(msg)
    elif (action_mode == "ACTION_LIFT_OUT"):
        liftOutFunction(msg)

def floatFunction():
    print ("agv_main.py-Do Nothing")
def manualFunction(msg):
    print ("agv_main.py-manualFunction")
    global pub_cmd_vel
    pub_cmd_vel.publish(msg)
    rospy.loginfo("agv_main.py-pub_cmd_vel")
def navigationFunction(msg):
    print ("agv_main.py-navigationFunction")
    global pub_move_base_goal
    pub_move_base_goal.publish(msg)
    rospy.loginfo("agv_main.py-pub_move_base_goal")
def initPoseFunction(msg):
    print ("agv_main.py-initPoseFunction")
    global pub_init_pose
    initPose = PoseWithCovarianceStamped()
    initPose.header = msg.header
    initPose.pose.pose = msg.pose
    pub_init_pose.publish(initPose)
def chargingInFunction(msg):
    print ("agv_main.py-chargingInFunction")
    global pub_charging
    pub_charging.publish(msg)
    rospy.loginfo("agv_main.py-publish chargingIn: " + str(msg))
def chargingOutFunction(msg):
    print ("agv_main.py-chargingOutFunction")
    global pub_charging
    pub_charging.publish(msg)
    rospy.loginfo("agv_main.py-publish chargingOut: " + str(msg))
def liftInFunction():
    print ("agv_main.py-liftInFunction")
def liftUpFunction(msg):
    print ("agv_main.py-liftUpFunction")
    global pub_lift
    pub_lift.publish(msg)
def liftDownFunction(msg):
    print ("agv_main.py-liftDownFunction")
    global pub_lift
    pub_lift.publish(msg)
def liftOutFunction():
    print ("agv_main.py-liftOutFunction")

if __name__ == '__main__':
    rospy.init_node('agv_main', log_level=rospy.DEBUG)

    rospy.Subscriber("agv_action", agv_action, initialAgvAction)
    rospy.loginfo("agv_main.py-Subscriber topic /agv_action")

    rospy.spin()